README.md
## ⚠️ Deprecation warning ⚠️
__This software is unlikely to be updated, mainly because the approach of relying on matplotlib for rendering proved to be very cumbersome and at the same time not powerful enough for raster effects.__

__Maybe I'll release a new library with a similar scope but rendering as raster, but don't hold your breath :/__

## Geoshiny


This is a tool to create highly customized maps from a PostGIS instance populated with Openstreetmap data.

__NOTE__: this software is under development, the interface and the functionalities can and probably will change.

## Features

* Generate customized maps of cities and small regions
* Outputs:
  * SVG
  * PNG
  * ...and many others
* Store a filtered intermediate representation in JSONL to easily generate images without a database

![example generated map](example.png)


## Input data

First you need a PostGIS database with the data to represent in flex format. To create that, get a PDB file of the region you are interested to represent (you can also import the whole planet.osm file) and use osm2pgsql with the flex output to run the [pgosm-flex](https://github.com/rustprooflabs/pgosm-flex) scripts.

If you have Docker, you can find a script that populates a PostGIS instance for you in `script_utils/database_from_pbf.sh`, run it with the pbf file name as an argument.

The connection string for this database has to be put in the environment variable `PGIS_CONN_STR`

    export PGIS_CONN_STR=postgres://username:password@localhost:5432/osm_data

## Installation

You will probably need to install the `geos` library on your system, on Ubuntu for example is the package `libgeos-dev`, on macOS using bres it's `geos`.
## Usage

NOTE: this is under development, usage will change soon

The library expects two callbacks, __representation__ and __renderer__.

The `representation` is called for each map feature in the given extent, receives the OSM tags as an input and returns `None` for features to be ignored or an arbitrary Python object (usually a dictionary) with the relevant information to represent. Can also alter the Shapely geometry.

The `renderer` will receive the output of the `representation` function and produce the matplotlib attributes like color and alpha.

So one takes care of deciding *what* to represent and the other of *how* to represent it. This decoupling allows to change representation and store intermediate values in a file.
Using `file_to_representation` you can generate the representation once and render different extents with different styles easily without even running a database instance.

```python
import asyncio

from geoshiny.database_extract import raw_data_from_extent
from geoshiny.types import (
    ExtentDegrees,
    Geometry2DStyle,
)

from geoshiny.draw_helpers import (
    data_to_representation,
    data_to_representation_file,
    file_to_representation,
    representation_to_figure,
)

def representation(osm_id: int, geom, tags: dict) -> Optional[dict]:
    if tags.get("bicycle") == "designated":
        return dict(path_type="bike")
    if "water" in tags:
        return dict(surface_type="water")

    if tags.get("landuse") == "grass":
        return dict(surface_type="grass")
    if tags.get("leisure") == "park":
        return dict(surface_type="grass")
    [...]

def renderer(osm_id: int, shape: BaseGeometry, d: dict):
    water_style = Geometry2DStyle(facecolor="blue", edgecolor="darkblue", linewidth=0.1)
    grass_style = Geometry2DStyle(facecolor="green", linewidth=0.1)

    if d.get("surface_type") == "water":
        return water_style
    [...]

# northern part of Rostock, Germany
extent = ExtentDegrees(
        latmin=54.0960,
        latmax=54.2046,
        lonmin=12.0029,
        lonmax=12.1989,
    )
# this directly renders to a file
generate_chart(
  'generated.png',
  extent,
  representation,
  renderer,
)
# but you can run the steps yourself
loop = asyncio.get_event_loop()
db_data = loop.run_until_complete(raw_data_from_extent(extent))
reprs = data_to_representation(db_data, entity_callback=representation)

db_img = representation_to_figure(reprs, extent, renderer, figsize=3000)

db_img.savefig("image.png")
db_img.savefig("image.svg")

data_to_representation_file(
        db_data,
        'somefile.jsonl',
        entity_callback=representation,
    )
img2 = representation_to_figure(
        file_to_representation('somefile.jsonl'),
        extent,
        renderer,
        figsize=3000,
    )
img2.savefig("image2.png")
img2.savefig("image2.svg")

```

## Testing

NOTE: this will also probably change, I'm looking at ways to run the tests without git-lfs

To run the test you need git-lfs and docker installed. If you didn't have git-lfs enabled before cloning the repo, you have to do `git lfs pull` to download the SQL dump file first.
Use `make test-from-zero` to perform a complete integration test, it will use a dump of a small postgis DB of around 200 MB to create a dockerized postgis instance and run scripts against that. The instance id deleted after the test or in case of errors.

# TODO

- [x] "Proper" automated tests with a reasonable data fixture
- [x] Examine the possibility of removing GDAL or making it optional, it's a pain to install
- [x] Add labeling/text
- [x] XKCD style output (from matplotlib, should work out of the box)
- [ ] layers/pipelines to further process the output
- [ ] Offer both async and sync access if possible, hiding the loop to sync users
- [ ] Visual comparison of output images (may require opencv as a test dependency, is it worth it?)
- [ ] Helper to generate world files (https://en.wikipedia.org/wiki/World_file)
- [ ] 3D output (check QGIS formats / glTF)
- [ ] tileset output?
- [ ] Create and document helpers to make the usage simpler (once the interface is stabilized)
- [ ] Examples and screenshot gallery
- [ ] Spatialite support?
- [ ] Optional pipeline definition interface to combine processing steps?
- [ ] Graph functionalities (e.g. show road distance from a set of POI) ?
- [ ] MoviePy integration example
- [ ] Integrated webserver/UI? Maybe better off as a separate thing


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.0.4]

### Changed
- Upgrade to Shapely 2.0
- Tolerate tables without a clear geometry type
- GDAL is now removed as a dependency, [the compatibility with Shapely was an issue](https://github.com/OSGeo/gdal/issues/3779)
- It's possible to add labels to the 2D chart

## 0.0.3
### Changed
- GDAL is now an optional dependency, under the `[geotiff]` extra
- A simple `generate_chart` helper allows the usage as a one-liner


geoshiny/__main__.py
import logging
from typing import Optional

from shapely.geometry.base import BaseGeometry

from geoshiny.types import (
    ExtentDegrees,
    Geometry2DStyle,
)
from geoshiny import generate_chart

logging.basicConfig(
    level=logging.DEBUG,
    format="%(asctime)s %(levelname)s: %(message)s",
)

logger = logging.getLogger(__name__)


def nice_representation(osm_id: int, geom, tags: dict) -> Optional[dict]:
    if tags.get("bicycle") == "designated":
        return dict(path_type="bike")
    if "water" in tags:
        return dict(surface_type="water")

    if tags.get("landuse") == "grass":
        return dict(surface_type="grass")
    if tags.get("leisure") == "park":
        return dict(surface_type="grass")
    if tags.get("natural") == "scrub":
        return dict(surface_type="wild grass")

    if "building" in tags and tags["building"] != "no":
        if "building:levels" not in tags:
            return dict(surface_type="building")
        else:
            try:
                level_num = float(tags["building:levels"])
            except ValueError:
                logger.debug(f"Invalid number of floors: {tags['building:levels']}")
                return dict(surface_type="building")
            return dict(surface_type="building", floors=level_num)
    return None


def nice_renderer(osm_id: int, shape: BaseGeometry, d: dict):
    water_style = Geometry2DStyle(facecolor="blue", edgecolor="darkblue", linewidth=0.1)
    wild_grass_style = Geometry2DStyle(
        facecolor="green",
        linewidth=0.1,
        min_label_area_ratio=0.002,
        label=dict(
            text="WILD grass",
            color="green",
            ha="center",
            va="center",
            bbox={"fc": "0.8", "pad": 0},
        ),
    )
    grass_style = Geometry2DStyle(
        facecolor="darkgreen", linewidth=0.1, label=dict(text="grass")
    )

    missing_levels = Geometry2DStyle(
        facecolor="red", edgecolor="darkred", linewidth=0.05
    )
    tall_build = Geometry2DStyle(facecolor="black", edgecolor="black", linewidth=0.05)
    low_build = Geometry2DStyle(facecolor="grey", edgecolor="darkgrey", linewidth=0.05)

    bike_path = Geometry2DStyle(linestyle="dashed", color="yellow", linewidth=0.1)

    if d.get("path_type") == "bike":
        return bike_path

    surface_type = d.get("surface_type")
    if surface_type == "building":
        if "floors" not in d:
            return missing_levels
        else:
            if d["floors"] > 2.0:
                return tall_build
            else:
                return low_build

    if surface_type == "grass":
        return grass_style
    if surface_type == "wild grass":
        return wild_grass_style
    if surface_type == "water":
        # from shapely import affinity
        # return water_style, affinity.rotate(shape, 90, origin='centroid')
        return water_style


if __name__ == "__main__":
    # # Most of Berlin
    # e = ExtentDegrees(
    #     latmin=52.4650,
    #     latmax=52.5805,
    #     lonmin=13.2911,
    #     lonmax=13.5249
    # )

    # # Berlin central station
    # extent = ExtentDegrees(
    #     latmin=52.5275,
    #     latmax=52.5356,
    #     lonmin=13.3613,
    #     lonmax=13.3768,
    # )
    # # small piece of Central Park in NY
    # extent = ExtentDegrees(
    #     latmin=40.78040,
    #     latmax=40.78280,
    #     lonmin=-73.96050,
    #     lonmax=-73.96350,
    # )

    # # Bicocca University main buildings, Milan, Italy
    # extent = ExtentDegrees(
    #     latmin=45.5098,
    #     latmax=45.5200,
    #     lonmin=9.2041,
    #     lonmax=9.2195,
    # )
    # northern part or Rostock, Germany
    extent = ExtentDegrees(
        latmin=54.0960,
        latmax=54.2046,
        lonmin=12.0029,
        lonmax=12.1989,
    ).enlarged(-0.8)

    # most of Berlin, takes:
    # 6 minutes to read all the data
    # 4 minutes to generate the figure
    # 2 minutes to generate the PNG
    # 2 minutes to generate the SVG
    # 1 minute to generate the geoTIFF
    # extent = ExtentDegrees(
    #     latmin=52.4215,
    #     latmax=52.6106,
    #     lonmin=13.1180,
    #     lonmax=13.6368,
    # )
    # import matplotlib.pyplot as plt

    # with plt.xkcd():
    generate_chart("generated.png", extent, nice_representation, nice_renderer)
    logger.info("done!")
